cli/azd/cmd/middleware/environment.go

```diff
@@ -0,0 +1,82 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/azure/azure-dev/cli/azd/cmd/actions"
+	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
+	"github.com/azure/azure-dev/cli/azd/pkg/lazy"
+)
+
+// EnvironmentMiddleware is a middleware that loads the environment when not readily available
+type EnvironmentMiddleware struct {
+	lazyAzdContext *lazy.Lazy[*azdcontext.AzdContext]
+	lazyEnvManager *lazy.Lazy[environment.Manager]
+	lazyEnv        *lazy.Lazy[*environment.Environment]
+	envFlags       internal.EnvFlag
+}
+
+// NewEnvironmentMiddleware creates a new instance of the EnvironmentMiddleware
+func NewEnvironmentMiddleware(
+	lazyAzdContext *lazy.Lazy[*azdcontext.AzdContext],
+	lazyEnvManager *lazy.Lazy[environment.Manager],
+	lazyEnv *lazy.Lazy[*environment.Environment],
+	envFlags internal.EnvFlag,
+) Middleware {
+	return &EnvironmentMiddleware{
+		lazyAzdContext: lazyAzdContext,
+		lazyEnvManager: lazyEnvManager,
+		lazyEnv:        lazyEnv,
+		envFlags:       envFlags,
+	}
+}
+
+// Run runs the EnvironmentMiddleware to load the environment when not readily available
+func (m *EnvironmentMiddleware) Run(ctx context.Context, next NextFn) (*actions.ActionResult, error) {
+	// We already have an environment, skip loading
+	// This will typically be the case when an environment has been created from a previous command like `azd init`
+	env, err := m.lazyEnv.GetValue()
```

Comment:
Do we know what's going on with #3973?

I think @vhvb1989 raised a few good points about the complexity here in coordination with middleware and lazy components all working together to achieve the somewhat straightforward effect of "Load environment based on flag/env var".

I'm also slightly worried about how we share the "lazy environment instance" across certain scoped lifetimes. This ma y just be an awareness thing since I'm not fully aware of the rules that would result in the behavior described by #3973. It could be unrelated completely.

I'd love a design where we just have commands call something like `loadEnvironment(flag)`. `azdContext` might be a perfect candidate for this. I'd want us to make things simple and have commands just load in from disk on-command, or if we do want to reshare state, we'd lay down clear rules on how that is managed so that we avoid future issues.

Reply:
Re #3973 - i will look into this, but any azd env vars set in a hook should then be immediately available when the context switches back to azd or if another hook script is fired afterwards.  If this is no longer the case then there is a bug here.

The lazy is just a wrapper around the lifetime of the environment scope.  Since azd environment is registered as scoped this instance will honor that and a new environment will be constructed if required in a new scope.  An example of this would be sub commands within an azd workflow.

Part of this change uses the registration of `Lazy[environment]` as the source of truth for all `environment` instances.  So when the instance is updated on this lazy instance here in the middleware that is ensuring that all downstream usages of the environment are using the same instance within the current scope.

I understand the want for commands to simply just call functions to do things like loadEnvironment or loadContext.  The issue I have is that as the code base grows over time this creates higher cognitive load in my opinion on knowing what are the correct order of operations a developer needs to do to get an instance of an environment.

While the middleware / IoC container may have higher initial cognitive load to understand what is happening behind the scenes it ensures that all consumers of components are getting the correct instance simply by having that component injected into their command/action/component, etc.

Reply:
FWIW, I do believe that just a factory function like `func loadEnv(*envFlag) (*environment.Environment, error)` is demanding the right amount of cognitive load from the developer. It shows how the environment should be constructed from without exposing the details of the construction. This also still allows the full flexibility of hiding away what the "correct instance" is.